sqlglot/parser.py

```diff
@@ -5475,6 +5476,53 @@ def _parse_comprehension(
             condition=condition,
         )
 
+    def _parse_heredoc(self) -> t.Optional[exp.Heredoc]:
+        tags = []
+
+        if self._match_text_seq("$"):
+            tags.append("$")
+        else:
+            return None
+
+        if self._curr and self._is_connected():
+            if not self._match_text_seq("$"):
+                tag_text = self._curr.text
+                tags.append(tag_text)
+                self._advance()
+            else:
+                tag_text = None
+                tags.append("$")
+        else:
+            self.raise_error("No closing $ found")
+
+        if tag_text:
+            if self._is_connected() and self._match_text_seq("$"):
+                tags.append("$")
+            else:
+                self.raise_error("No closing $ found")
+
+        heredoc_start = self._curr
+
+        while self._curr:
+            index = self._index
```

Comment:
shouldn't this index be outside of the for loop? perhaps it should be at line 5480?

Reply:
since I'm not using `_match_text_seq`, the `index` keeps track of `self._curr` is before we try to match subsequent tokens. If the matching fails at any point, we return to `index` before calling `self._advance()` to advance `self._curr` (`index` here serves a similar purpose to `index` in `_match_text_seq`)